.circleci/config.yml

```
@@ -1,5 +1,82 @@
-version: 2.0
+version: 2.1
+commands:
+  save_deps_cache:
+    steps:
+      - save_cache:
+          key: v1-mix-deps-cache-{{ checksum "mix.lock" }}
+          paths: ["deps"]
+  restore_deps_cache:
+    steps:
+      - restore_cache:
+          keys:
+            - v1-mix-deps-cache-{{ checksum "mix.lock" }}
+            - v1-mix-deps-cache-
+  workspace_init:
+    steps:
+      - checkout
+      - restore_deps_cache
+      - run: mix deps.clean --unused
+      - run: mix deps.get
+      - save_deps_cache
+  save_build_cache:
+    parameters:
+      env:
+        description: Mix environment
+        type: string
+        default: dev
+    steps:
+      - save_cache:
+          key: v1-build-cache-{{ arch }}-<< parameters.env >>-{{ checksum "mix.lock" }}
```

I believe the build cache key should also include the branch name. My understanding is that in the current setup, two branches could repeatedly override each other's cache. It probably doesn't break anything given that `mix compile` is always performed, but it doesn't seem correct either.
```suggestion
          key: v1-build-cache-{{ arch }}-{{ .Branch }}-<< parameters.env >>-{{ checksum "mix.lock" }}
```

---

> could repeatedly override each other's cache
They won't as CircleCI only updates the cache if one with the set key doesn't exist, so only the first one will be stored and all the following builds will use that.

This means the only difference it would make is that after creating a branch, all the following builds would use the first build of that branch instead of the cache used on master at the moment of branching, so the code might be a bit more fresh, but still, the branch assumes the code will change, so there won't be much improvement from updating cache after creating a new branch. The solution would be to include a commit hash or even better hash of the contents of `lib` and/or `test` dirs, but again - the difference will be unnoticeable.

---

Anyway, any suggestions regarding CI should be moved to `membrane_template_plugin` this is only a copy-paste from there